lib/tree.rb

```diff
@@ -16,51 +16,237 @@ def initialize
+  # Space Complexity: O(n) - Assuming the same array is passed down memory, simplified from O(2n).
+  # The array is accounts for a hash created for each node.
+  # basically the same as inorder. The stack size will be the height of the tree, which doesn't exceed O(n)
+  def preorder_helper(node, array)
+    if node
+      array << {:key=> node.key, :value=> node.value}
+      preorder_helper(node.left, array)
+      preorder_helper(node.right, array)
+    end
+
+    return array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
+  end
+
+  # Time Complexity: O(n) -- you only visit each node once. 
+  # Space Complexity: O(n) Assuming the same array is passed down memory, simplified from O(2n).
+  # The stack accounts for visiting every node once; the array is accounts for a hash created for each node.
+  # basically the same as inorder. The stack size will be the height of the tree, which doesn't exceed O(n)
+  def postorder_helper(node, array)
+    if node
+      postorder_helper(node.left, array)
+      postorder_helper(node.right, array)
+      array << {:key=> node.key, :value=> node.value}
+    end
+
+    return array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
+  end
+
+  # Time Complexity: O(n) - You have to visit every node once, even though you are only actually counting the number of levels 
+  # in the tree and not the nodes themselves, to confirm the height of each sub-tree. 
+
+  # Space Complexity: O(n) - Since you are checking every node per call/recursive call, each node gets its own 
+  # entry in the stack. 
+  def height_helper(node)
+    if node
+      left = 1 + height_helper(node.left)
+      right = 1 + height_helper(node.right)
+
+      return left > right ? left : right
+    else 
+      return 0
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def height
-    raise NotImplementedError
+    return height_helper(@root)
   end
 
   # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) ...? If we pretend the queue doesn't have an O(n) every time an element gets popped, O(n) -- all elements
+  # are operated on at most thrice to enqueue, pop, and load element into the return array. My guess is that even with 
+  # adding to the end of the array is always an O(n) operation, but the the shifting would be O(n^2) given that a balanced
+  # tree would have 2 nodes pushed in for every node pushed out, so every loop you are shifting an array with one element 
+  # more than the last. I just made a Queue object to be on the safe side. 
+
+  # Space Complexity: O(n) for two arrays that have max of length O(n) for all function calls -- one being the return array containing
+  # all the elements, the other being the queue
   def bfs
```

Comment:
👍 Correct on time/space complexity.  Nice work getting BFS to work with a queue.